pkg/controllers/metrics/node/counter/counter.go

```
@@ -0,0 +1,162 @@
+	}
+}
+
+func handleUpdatedNode(ctx context.Context, name types.NamespacedName, currLabels prometheus.Labels, pastLabels prometheus.Labels) {
+	// Only report node updates that affect tracked dimensions.
+	if reflect.DeepEqual(currLabels, pastLabels) {
+		return
+	}
+
+	prometheusLabelsFor[name] = currLabels
+
+	if err := decrementNodeCount(pastLabels); err != nil {
+		logging.FromContext(ctx).Warnf("Failed to decrement previous count for updated node [labels=%s]: error=%s", pastLabels, err.Error())
+	}
+	if err := incrementNodeCount(currLabels); err != nil {
+		logging.FromContext(ctx).Warnf("Failed to increment current count for updated node [labels=%s]: error=%s", currLabels, err.Error())
+	}
+}
+
+func getLabels(node *v1.Node) prometheus.Labels {
+	labels := make(prometheus.Labels)
+	if node != nil {
+		labels[metrics.ProvisionerLabel] = node.Labels["karpenter.sh/provisioner-name"]
+		labels[metricArchLabel] = node.Labels["kubernetes.io/arch"]
+		labels[metricInstanceTypeLabel] = node.Labels["node.kubernetes.io/instance-type"]
+		labels[metricTopologyRegionLabel] = node.Labels["topology.kubernetes.io/region"]
+		labels[metricTopologyZoneLabel] = node.Labels["topology.kubernetes.io/zone"]
+		labels[metricOsLabel] = node.Labels["kubernetes.io/os"]
+
+		for _, c := range node.Status.Conditions {
+			labelName, found := conditionTypeToMetricLabel[c.Type]
+			if found {
+				labels[labelName] = strings.ToLower(string(c.Status))
+			}
+		}
+	}
+
+	return labels
+}
+
+func incrementNodeCount(labels prometheus.Labels) error {
+	return updateNodeCount(labels, prometheus.Gauge.Inc)
+}
+
+func decrementNodeCount(labels prometheus.Labels) error {
+	return updateNodeCount(labels, prometheus.Gauge.Dec)
+}
+
+func updateNodeCount(labels prometheus.Labels, update func(prometheus.Gauge)) error {
+	nodeCount, err := nodeCountGaugeVec.GetMetricWith(labels)
```

optional:  Seems kind of backwards from the normal go idioms of failing early if err and happy case is not indented.